enhancements/machine-api/control-plane-machine-set.md

```diff
@@ -0,0 +1,908 @@
+    ReadyReplicas int32 `json:"readyReplicas,omitempty"`
+
+    // UpdatedReplicas is the number of non-terminated Control Plane Machines
+    // created by the ControlPlaneMachineSet controller that have the desired
+    // provider spec.
+    UpdatedReplicas int32 `json:"updatedReplicas,omitempty"`
+
+    // UnavailableReplicas is the number of Control Plane Machines that are
+    // still required before the ControlPlaneMachineSet reaches the desired
+    // available capacity. When this value is non-zero, the number of
+    // ReadyReplicas is less than the desired Replicas.
+    UnavailableReplicas int32 `json:"unavailableReplicas,omitempty"`
+
+    // Conditions represents the observations of the ControlPlaneMachineSet's current state.
+    // Known .status.conditions.type are: (TODO)
+    // TODO: Identify different condition types/reasons that will be needed.
+    // +patchMergeKey=type
+    // +patchStrategy=merge
+    // +listType=map
+    // +listMapKey=type
+    // +optional
+    Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+```
+
+### Implementation Details/Notes/Constraints
+
+The ControlPlaneMachineSet controller aims to act similarly to the Kubernetes StatefulSet controller.
+It will take the desired configuration, the `ProviderSpec`, and ensure that an appropriate number of Machines exist
+within the cluster which match this specification.
+
+It will be introduced as a new second level operator so that, if there are issues with its operation, it may report
+this via a `ClusterOperator` and prevent upgrades or further disruption until the issues have been rectified.
+Due to the nature of the actions performed by this controller, we believe it is important that it have its own
+`ClusterOperator` in which to report its current status.
+
+The `ControlPlaneMachineSet` CRD will be limited to a singleton within a standard OpenShift HA cluster, the only
+allowed name will be `cluster`. This matches other high level CRD concepts such as the `Infrastructure` object.
+Due to this limitation, the CRD will be cluster scoped.
+
+The behaviour of such a controller is complex, and as such, various features of the controller and scenarios are
+outlined in the details below.
+
+#### Desired number of Machines
+
+At present, the only source of the installed control plane size within OpenShift clusters exists within the
+`cluster-config-v1` ConfigMap in the `kube-system` namespace.
+This ConfigMap has been deprecated for some time and as such, should not be used for scale in new projects.
+
+Due to this limitation, we will need to have a desired replicas field within the ControlPlaneMachineSet controller.
+As we are not planning to tackle horizontal scaling of the control plane initially, we will implement a validating
+webhook to deny changes to this value once set.
+
+For new clusters, the installer will create (timeline TBD) the `ControlPlaneMachineSet` resource and will set the value
```

Comment:
The installer will continue to create the `Machine` resources for the control plane in addition to the `ControlPlaneMachineSet`, right?

Reply:
Yes that was the plan, for now, when the `ControlPlaneMachineSet` is created, it is expected that the Machines will already exist, it will go degraded if there are no Control Plane Machines. The CPMS will then take over management of the Machines that the installer creates